test/integration/filters/hanging_filter.cc

```diff
@@ -0,0 +1,57 @@
+#include <string>
+
+#include "envoy/registry/registry.h"
+
+#include "common/protobuf/utility.h"
+
+#include "extensions/filters/http/common/factory_base.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+#include "test/test_common/test_time_system.h"
+
+namespace Envoy {
+
+// A test filter that hangs unendingly during the decode/encode phase. This filter simulates a
+// slow/hanging filter that causes the thread executing the filter callbacks to block.
+class HangingFilter : public Http::PassThroughFilter {
+public:
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) {
+    mutex_.Lock();
+    while (true) {
+      cond_var_.Wait(&mutex_);
```

Comment:
my apologies; I should have explicitly suggested cond-var, but it would be more concise to use a wrapper in https://github.com/abseil/abseil-cpp/blob/master/absl/synchronization/notification.h , rather than condvar directly.

But the functional issue here is that the tests look like they will hang or leak. CI also seems unhappy at this time. You need to provide a mechanism to unblock the condvar by calling notification.Notify().

The tricky part here is plumb through access to this filter's API to the test. I don't know if there's a good pattern for that in Envoy. Once available you can unblock the hanging filter in the test and let them clean up.

Reply:
Got it, makes sense.